### ISSUE TITLE
Implement LAN Multiplayer System And Mode

### ISSUE THREAD

**Issue by J-Shep** [CorsixTH/CorsixTH#386]
I'm going to try to implement this if no one is working on this already.

**Comment by J-Shep** [comment]
I will aim to have a design done for this task by February 1st and when I've finished a design I will post it here.

**Comment by J-Shep** [comment]
>> I will aim to have a design done for this task by February 1st and when I've finished a design I will post it here.

I'm going to need another week to design what I'm going to implement for this task because I've been playing PC games this week in my free time at home instead of working on my CorsixTH tasks including my task of designing what I'm going to do for this issue.

If I've not been committed enough to this task to have at least an acceptable design done for it in the near future then I wouldn't mind passing on this task to a CorsixTH developer who would be more dedicated to working on it and getting it finished asap.

**Comment by J-Shep** [comment]
I will spend all of my CorsixTH development time tomorrow designing what I'm going to implement for my phase one release and then I will probably post at least the start of a design here by Monday.

**Comment by J-Shep** [comment]
I expect to finish implementing #445 soon, once this important task is done I will finish designing what I'm going to do for this task. 

I'm sorry I've not worked on this task for a long time despite this being a feature which all players must being looking forward to: the other tasks I've worked on instead will make implementing anything for CorsixTH and debugging its errors more easier.

**Comment by Cach54** [comment]
Is somebody still working on Multiplayer ?

I am new to CorsixTH but I could probably help a bit, I am a Web Project Manager/Developper.

**Comment by J-Shep** [comment]
@Cach54 I'm sorry to say that I've never actually started working on this despite saying that I would and making players excited including the player who sent me an email to thank me. I've worked on other things instead to help other developers find and fix problems, like implementing DBGp debugger support because I felt these tasks were more important to get done and would help to make the task of implementing multiplayer support easier.

I don't mind if someone who's a better more committed CorsixTH developer than me, who unlike me can be trusted to actually work on the things they promise to work on, wants to take over the responsibility of implementing multiplayer support.

I will work on this if no one else does, after I've finished implementing automatic error reports. If anyone starts working on this please post a comment here so that I will know not to spend time working on it myself.

If I ever decide to not work on this task or any other I will post a comment to announce this so that someone else can take it over.

**Comment by J-Shep** [comment]
@Cach54 I'm still in favor of using a design driven approach for this task, so even if you have no programming experience you could help me decide how it should be implemented in the design stage without having to be able to make programming level implementation decisions. We could use UML diagrams.

I'm going to make a promise which I'm going to try to keep: I will start working on a design by Monday and by the following Sunday I will have a finished design/ at some ideas to post here for @Cach54 and CorsixTH developers to critique.

**Comment by Cach54** [comment]
@J-Shep That is pretty awesome.

I actually have programming experience, I program almost everyday. But I am a web developper, PHP most of the time. I have C basis but I never heard about LUA. Which means I can implement functions if needed, but I would not be able at the moment to design or to take choice on architecture.

I already followed this repo, which means I am notified when there is activity.
I am looking forward to your designs.

Good luck.

**Comment by J-Shep** [comment]
@Cach54 When what will be implemented has been designed and agreed on in this issue discussion, I will make a programming task todo list so that you and other programmers could volunteer to do some of these tasks so that this programming work can be shared and multiplayer support can be implemented sooner.

When I created this issue discussion I intended to make a multiplayer system which would support a multiplayer single scenario mode like the singleplayer scenario mode: in which players would compete to win a scenario by achieving its objectives before the other players / having a better hospital score after X years had passed. 

So each player would have their own town map and they wouldn't have to be able to see their opponent's hospitals like they would in TH's shared town map multiplayer mode, so very little game state data would have to be shared between players and I don't think lag would be a coding issue for making this multiplayer mode playable over the internet. Support for allowing players to see their opponent's hospitals could be added later after the simpler implementation of this mode had been done.

Does anyone think the first multiplayer mode to be implemented should be the shared town map mode from TH instead? it would be much more complicated to implement.

**Comment by J-Shep** [comment]
>> @Cach54 When what will be implemented has been designed and agreed on in this issue discussion, I will make a programming task todo list so that you and other programmers could volunteer to do some of these tasks so that this programming work can be shared and multiplayer support can be implemented sooner.

I think it would be awesome to have lots of CorsixTH programmers working together on getting multiplayer support implemented asap by each working on a task from this todo list and sharing a "multiplayer" repository branch which @TheCycoONE could create in the CorsixTH repository. 

Though with lots of developers working together someone would need to manage everyone to avoid teamwork coordination problems & conflicts. I would be a terrible project manager so someone else would have to take on this role. @Cach54 @lbakyl @Lego3 Being experienced and professional software project managers do any of you guys think you could have time to help with this?

**Comment by J-Shep** [comment]
>> I'm going to make a promise which I'm going to try to keep: I will start working on a design by Monday and by the following Sunday I will have a finished design/ at least some ideas to post here...

 I've been too busy this week to spend much time working on this so I will need more time to work on this before I post ideas/a design here. I might be ready to post ideas/ a design here on Friday.

**Comment by Alberth289346** [comment]
I don't think you'll have to worry about project management, so far there aren't enough devs around to cause any problem :)

As for making progress, I don't know what kind of design you have in mind. If you have a good idea in mind of what you want, great! Just ignore me :)
If you are struggling to find a good form, allow me to suggest one. If you don't like it, again, feel free to ignore me. Otherwise, read it, think about it, and use what you see as useful.

A multi-player system is a big addition. From what I read here, I get a picture of a feature list, and a desire to make a proper design. (You're talking design diagrams of a finished system at some point above, not sure if that's still your plan.) Design diagrams are great, but they feel a like a bit too much for now, to me. If you make a diagram, you have to make a lot of decisions, where many of them are pretty much unclear. As a result, it's hard to make a diagram. (If I am wrong, you can stop reading here :) )


An alternative can be to make a design but with much less details. Maybe you could start by describing how a multi-player game works, mostly in words? Assume each player/computer is a circle, have lines representing the connections. That's the system. At this point, don't worry what each circle does exactly, only worry about the overview picture, how do these circles play a game together? What information is sent over the lines?? What is computed in each circle?? How do you keep time in sync? (all having the same time in the game?) In other words, what do you communicate over the lines?

Think about a page of text, probably less. In other words, very short, not much detail.

If there are several solutions, you can describe all of them, or point out differences (hopefully not too many or it gets too complicated to understand).


As a next step from the communication patterns above, you can make a check for the features that you want. Basically, assume someone implemented the circles and lines system above. Now how do you provide the wanted features for each circle? What information needs to be communicated between who?
This works in two ways. At the one side, you get an idea of communication needs and patterns. At the other side, you roughly check whether you got all information at the right spot when needed.

Then, you can consider what messages to send over the network in more detail, and how to handle processing and communication inside the circles.

**Comment by J-Shep** [comment]
@Alberth289346 I'm sorry for my late reply, I've read every word you've taken the time to write in your helpful comment.

>> If you are struggling to find a good form, allow me to suggest one. If you don't like it, again, feel free to ignore me. Otherwise, read it, think about it, and use what you see as useful.

I've not been making progress with this task and my other CorsixTH tasks because of being busy doing other things and because of a lack of dedication to spending some of my free time contributing to CorsixTH. My lack of dedication hasn't been helped by having a hacker interfere with my CorsixTH work in the past and possibly recently by creating unfixable errors to waste my time. I do enjoy programming but I don't enjoy trying to fix unfixable compile & runtime errors. I know a hacker did this at least once in the past because they contacted me afterwards.

I agree that I should start with a high level conceptual design before trying to make a more detailed implementation design. I will consider using the diagram method you've described when I spend  tomorrow designing how I think multiplayer support should work. I will probably post a design here by Sunday.

Thanks for kindly trying to help me make progress with this task. I think seeing the long post you've taken the time to write in order to help me, might motivate me to become a more committed and reliable CorsixTH contributor than I've been thus far this year.

**Comment by J-Shep** [comment]
I've started making a design document using Google Docs and I will make it public tonight or tomorrow.

**Comment by J-Shep** [comment]
>> I've started making a design document using Google Docs and I will make it public tonight or tomorrow.

I'm not going to post a link to it here today now because I still have a lot of work I want to do for it before anyone takes the time to read it and makes comments about it.

**Comment by J-Shep** [comment]
I will do a bit more work on my design document this weekend and then I will post a link to it here but it won't be finished by then. 

I then want to focus on getting my "Automatic error reports" feature implemented for 0.50 before its July 12th feature freeze deadline. So I may not resume designing this until after July 12th.

**Comment by zsolmanz** [comment]
> A. Design the user interface layout for a "Available LAN Servers" screen, my requirements:
> B. Design the user interface layout for a LAN game's pre-launch room, my requirements:

Hey @J-Shep, do you still want these done?

I just knocked up a super basic main menu 'multiplayer' button addition to test out libraoffice (to be honest, I found MS paint to be easier), you can find it [**here**](https://drive.google.com/open?id=0BzMXwT_4ohHBOWxoUUtPQ2JQM1E).

Biggest issue I came up with is:  What font does Theme Hospital (or Corsix) use as the main menu font?  I had to manufacture the word 'multiplayer' in the above picture from individual letters (hence no capital M), but that'd be a pain for the more complicated designs you've asked for.

**Comment by MarkL1961** [comment]
New text is added to the English file and then translated into other languages by the translators.  Look at line 244 in C:\CorsixTH 0.51\Lua\languages\English.lua for an example of Load Game in the same menu

**Comment by zsolmanz** [comment]
Unfortunately the TH font file is a .dat or a .tab that I can't yet open - does anyone know the name of the font?

**Comment by J-Shep** [comment]
>> A. Design the user interface layout for a "Available LAN Servers" screen
B. Design the user interface layout for a LAN game's pre-launch room 

Hi @zsolmanz thanks for offering to help by designing what these dialogues will look like.

I think including layouts designs for these new UI windows in the detailed design document I'm making would be worthwhile, so if you would like to design these windows for me then be my guest. 

You don't need to create mockup images depicting exactly what you think these windows should look like, just having basic black & white UI layout boxes drawings would suffice:
![Example UI layout drawing](http://www.localwisdom.com/wp-content/uploads/2013/08/LinearLayout.png)

So you could just make layout drawings with text boxes for window components in Libre Office Draw or [Google Drawings](https://docs.google.com/drawings/d/117Xa2d1vAY-0JX_S2sCCsooYDvm2AgvoXgdU9tnn1V0/edit).

I would prefer you to make these layout drawings with [Google Drawings](https://docs.google.com/drawings/d/1A-qEgOwFXAhYR_Ut3PAtVy7LIAAHhfFWq79mX02-6UM/edit) because I'm making the design document in Google Docs, so I could import your "Google Drawings" diagrams in an editable way and then you could edit them in this document by just visiting it in a web browser with Google Docs. Or this document could have links to your drawings in your Google web space where anyone with a web browser could view them and edit them in "suggestions" editing mode (if you've permitted them to do this) allowing you to accept or decline the suggested changes.

As well as taking you a lot less time to make, the text box components in these simple layout diagrams could quickly & easily be rearranged and reshaped. So if you do want to design these windows I think you should use the quick & malleable design drawing method.

I would recommend you try out [GIMP](http://www.gimp.org/) and [learn how its layers feature works] (https://www.youtube.com/watch?v=xUMiHZScthE). I've always used it to create CorsixTH dialog mockup images:
* I have a background image layer, like the view of a hospital if I'm creating a mockup image of an in game window.
* Above this I have a layer depicting the Window's background with its borders.
* And then I have layers for each image of a UI component & for groups of components I want to edit & move together.

So I would have a layer for a button so that I could move the image of it around my drawing by simply selecting its layer and then dragging and dropping it & so that I could edit edit the image of the button without makes changes to what the other layers depict.

My CorsixTH contribution priority right now is to help fix bugs for 0.50 because players will be able to benefit from this work when this next version is released where as they won't benefit from any of my work for multiplayer support until 0.60/0.70/0.80. 

I intend to resume my work on my design document soon though and when I think its ready to become public so that other developers can review my design ideas, I will post a link to it here.

**Comment by J-Shep** [comment]
>> Then in C:\CorsixTH 0.51\Lua\dialogs\resizables\main_menu.lua to see how the buttons are added.
I think the game will use whatever font is specified here from either the TH font file or a user font if they have specified one i.e. where none English characters are used

There will be an optional and much simpler method for programming UI dialogs, if my implementation of Layout Managers is accepted. Based on Java's layout managers, I've implemented simple layout managers to help me program the new dialogs for automatic error reports.

**Comment by zsolmanz** [comment]
Are [**these**](https://drive.google.com/open?id=0BzMXwT_4ohHBM1B5NVJ5MklTNVU) any good?

**Comment by J-Shep** [comment]
I think a good minimum size these windows should support is 1024x768 because I think all netbook screens support this size as a supported screen resolution.

Having information like player pings displayed in game lobbies and a button for filtering listed servers wouldn't be necessary for LAN multiplayer but by including UI information and components which would be useful for internet games, these LAN windows would be ready to be reused for internet multiplayer games.

**The chat box section of each window**
* The fax machine TH image is pretty cool imo so I can appreciate your idea of using it as a chat history background below the chat message input box, with players being able to use its button to control who they chat with in a game/servers lobby if they don't want to chat with everyone. If it will look eye pleasing  rather than intrusive amongst all the other UI components then I would personally be in favour of using this image in this way.

**The multiplayer/servers lobby window**
![image](https://cloud.githubusercontent.com/assets/6353342/8827012/f595694a-3081-11e5-8cdb-994a8405ff23.png)
* For use as a LAN & internet multiplayer servers window, I personally  think all the information table columns you've chosen would be worth having though the player name column isn't essential and so could be removed to save screen space on smaller screens.
* If there will be a "Player Name" column I personally think "Host" would be a better name for it.
* The options button: players could use the main menu settings window to change their multiplayer name and then this multiplayer button could be named "Filters".
* One UI component your design doesn't have to go along with your inclusion of chat UI components in this window, is a list of the players who are available to chat with.
* A missing button which would be essential for internet gaming is a friends list button which when clicked could display a friends list window with UI components for adding & removing players from this list. It could be located in the chat section of each window you've designed

**The game lobby window**
* I think a host should be prompted to choose a game name after clicking the "host game" button and before the game lobby window is displayed but being able to change a game's name in its lobby window could be useful.
* I think there will end up being at least two multiplayer gameplay modes game hosts will be able to choose to host games for: the multiplayer mode from TH where players share a town map and a multiplayer single scenario mode which would work like its singleplayer counterpart except that a player would be competing with other players' hospitals instead of bot hospitals. The multiplayer single scenario mode is what I intend to design and implement first because I think it would be easier to implement as both a LAN & internet playable mode.
*  In a 1024x768 sized game lobby window I personally think having the connected players info table take up as much vertical space as you've planned would make it take up a lot more space than necessary and this space could instead be used for providing a description of the selected scenario along with an image of its map.

In case it provides inspiration here's a mockup image I once created in GIMP for a layout idea I had for the singleplayer single scenario window, I've lost its GIMP source file:
![Single Scenario Window Mockup](https://www.mediafire.com/convkey/277f/ewzkevjgv8m1gs16g.jpg)

**Comment by J-Shep** [comment]
I've been responsible for working on this task for nearly a year now and regrettably I've not gotten around to actually properly starting work on this mainly because when I've had free time to spend doing programming, I've distracted myself with other CorsixTH tasks and programming tasks for other free games. 

From now on I'm going to have a lot less free time and I want to spend more of my free time doing other hobbies besides programming things for free games. 

So I've decided that I no longer want to have the big responsibility of being the CorsixTH programmer responsible for working on this time consuming difficult task which all CorsixTH players would want finished asap.

I've not made any code for this task and I don't want to give the design document I worked on to the developer who will take over this task because my design document is rubbish.

**Comment by TheCycoONE** [comment]
Thanks for the update @J-Shep, good luck in your future endeavors.

**Comment by Alberth289346** [comment]
Good luck from me too

**Comment by J-Shep** [comment]
>> Thanks for the update @J-Shep, good luck in your future endeavors.

>> Good luck from me too

@TheCycoONE  @Alberth289346 Thanks guys

**Comment by lewri** [comment]
I think we can agree that this is actually a post 1.0 goal? As such, I should probably remove it from that project.

**Comment by ARGAMX** [comment]
> I think we can agree that this is actually a post 1.0 goal? As such, I should probably remove it from that project.

I fully support this.
The multiplayer is clearly not what made this game what it is for us.  
Therefore, it’s not part of the core set. 
Therefore, it shouldn’t be part of 1.0.
There’s also a concern that the multiplayer might never appear here at all.
I wouldn’t want the release of 1.0 to be indefinitely delayed because of this.
